### ISSUE TITLE
Missing root cert x5c chain fails to verify

### ISSUE THREAD

**Issue by ptoffy** [apple/swift-certificates#199]
Hi! So I've been using swift-certificates in JWTKit to verify x5c chains in token headers. I'm trying to validate this token in particular:
```
eyJhbGciOiJFUzI1NiIsInR5cCI6IkpXVCIsIng1YyI6WyJNSUlCanpDQ0FUVUNGSDhFWFBJbDBRbDQwYzdKOEhkK2R6QWgwY3dVTUFvR0NDcUdTTTQ5QkFNQ01FNHhDekFKQmdOVkJBWVRBbFZUTVJNd0VRWURWUVFJREFwVGIyMWxMVk4wWVhSbE1RNHdEQVlEVlFRS0RBVldZWEJ2Y2pFYU1CZ0dBMVVFQXd3UmFXNTBaWEp0WldScFlYUmxMV05sY25Rd0hoY05Nak14TURJME1UTTBNalU1V2hjTk1qUXhNREl6TVRNME1qVTVXakJHTVFzd0NRWURWUVFHRXdKVlV6RVRNQkVHQTFVRUNBd0tVMjl0WlMxVGRHRjBaVEVPTUF3R0ExVUVDZ3dGVm1Gd2IzSXhFakFRQmdOVkJBTU1DV3hsWVdZdFkyVnlkREJaTUJNR0J5cUdTTTQ5QWdFR0NDcUdTTTQ5QXdFSEEwSUFCRU85VjZLQWhSR0l4QWx2Ukl6U3BtSVRmVVZiWHl5ZGMvZWdlbHpLLzZ3NDEySmc4RDJlSVRkOHVDRmxnVmh4WUlkR1pNN1hYUWhaNmhOZnE2S3JyUG93Q2dZSUtvWkl6ajBFQXdJRFNBQXdSUUloQUxJaXhudE93V3o4NlFHa0g0SGNnT09malBiczlBUVpXYm1HYkpKRjRWRWZBaUI4ZUVLRi9WQllvWVhRREQzVHpLNUlEMkdzbXplMzhpNk56ek9ndHRMam9nPT0iLCJNSUlCNXpDQ0FZeWdBd0lCQWdJVUw2ZnJzdHdldjdZTWtPWVNuUHVlSlQyR3g1UXdDZ1lJS29aSXpqMEVBd0l3UmpFTE1Ba0dBMVVFQmhNQ1ZWTXhFekFSQmdOVkJBZ01DbE52YldVdFUzUmhkR1V4RGpBTUJnTlZCQW9NQlZaaGNHOXlNUkl3RUFZRFZRUUREQWx5YjI5MExXTmxjblF3SGhjTk1qTXhNREkwTVRNME1qTTFXaGNOTWpneE1ESXlNVE0wTWpNMVdqQk9NUXN3Q1FZRFZRUUdFd0pWVXpFVE1CRUdBMVVFQ0F3S1UyOXRaUzFUZEdGMFpURU9NQXdHQTFVRUNnd0ZWbUZ3YjNJeEdqQVlCZ05WQkFNTUVXbHVkR1Z5YldWa2FXRjBaUzFqWlhKME1Ga3dFd1lIS29aSXpqMENBUVlJS29aSXpqMERBUWNEUWdBRWt6WGFFQTlDSVpyUkVmQ0Mra05tM0pxZDdmR0ZFT05Ia2p5TFJ4NS83SUFYeHB4TGZ0WWNoOTU1K1VRNVhHOHdUZ2tNQ0NaNG9LRjhNMXg3Zkw3cXU2TlFNRTR3REFZRFZSMFRCQVV3QXdFQi96QWRCZ05WSFE0RUZnUVVyRU53VW02VDBSbmUxTW9MY3lWQ2NhVTVvTTB3SHdZRFZSMGpCQmd3Rm9BVVkxM3oxRVhHbjlVQmZwZ3BmVTl0UWJ2TVJRZ3dDZ1lJS29aSXpqMEVBd0lEU1FBd1JnSWhBSW04bStLb0RFbktBdUgrZUZROGJWSDJkc3p2NlcveCtwNE9zZERzd0VrNUFpRUF5bWd1SmdxQUpZU3NDdzdYM0pDVVBNY29LdGFRRzZNamhRdThrWlpCQUNJPSJdfQ.eyJjb29sIjp0cnVlfQ.d7AmnXaDUu1eU5ufBQ5ruailzBbJVud3RtqAcDYApvP6fPC1SttYmDKbsjUHx4q-NEt14n6xZ0Np_zy-OtM6lw
```
this is a JWT with an x5c chain which is missing a root certificate, it should however still be verifiable, in fact [jwt.io](https://jwt.io) does mark it as verified. swift-certificates however fails to verify it, returning `.couldNotValidate(failures)` with an empty `failures` array. Here's the code I'm using: https://github.com/vapor/jwt-kit/blob/main/Sources/JWTKit/X5C/X5CVerifier.swift#L165-L192. Is there something wrong with my code or is this an issue with the library? Thanks!

**Comment by Lukasa** [comment]
Could not validate with an empty failures array usually indicates a failure in chain building.

If you want to understand what the library is doing as part of chain building, you can add a [diagnostic callback](https://swiftpackageindex.com/apple/swift-certificates/1.2.0/documentation/x509/verifier/validate(leafcertificate:intermediates:diagnosticcallback:)). This will give you an object that can either be interpolated into logs using its `description` or printed using its [`multilineDescription`](https://swiftpackageindex.com/apple/swift-certificates/1.2.0/documentation/x509/verificationdiagnostic/multilinedescription) representation to explain what we're doing.

Can you put the output from that diagnostic here?

As an FYI, we generally don't recommend enabling these logs consistently as they will markedly slow down chain building, which can be a serious performance bottleneck even without the diagnostic logs. However, if you're running a service it can be useful to have a diagnostic callback that invokes a swift-log logger at debug level, and then offer configuration to boost that specific logger in some circumstances.

**Comment by Lukasa** [comment]
It would also likely be useful if you could supply the root certs you're using as well.

**Comment by ptoffy** [comment]
Yep you're right, after debugging with the diagnostics callback I realised [here](https://github.com/vapor/jwt-kit/blob/main/Sources/JWTKit/X5C/X5CVerifier.swift#L170) I'm doing 
```swift
// Setup an untrusted chain using the intermediate certificates
let untrustedChain = CertificateStore(certificates.dropFirst().dropLast())
```
which makes sense when there's a root but removes an intermediate if no root is in the chain. How should this best be handled? Because _not_ removing it would mean leaving the root in the `untrustedChain` which doesn't really make sense? Maybe something like
```swift
let untrustedChain: CertificateStore
if certificates.count > 1 {
    let potentialIntermediates = certificates.dropFirst()
    let intermediates = potentialIntermediates.filter {
        $0.subject != $0.issuer
    }
    untrustedChain = CertificateStore(intermediates)
} else {
    untrustedChain = CertificateStore()
}
```

**Comment by Lukasa** [comment]
You can leave the root in the untrusted chain. It doesn't help, but neither does it hurt. Our verifier has the following logic:

1. We won't trust a chain that doesn't end in a trusted root.
2. We'll try to build through the roots before we try to build through the intermediates.
3. We won't build a long chain where a shorter equivalent one exists.
4. We won't include the same cert twice.

The result of this is that having the root in the chain is simply ignored: we'll only ever build through it if it is different in some way than the root in the trusted roots, e.g. if it's cross-signed. So you can just avoid the filter altogether.

**Comment by Lukasa** [comment]
For that matter, you _can_ also leave the leaf in the untrusted chain. Again, it doesn't help, but neither does it hurt. We'll burn a little bit of extra CPU hashing the distinguished name, so it's nice of you to remove it if it's easy to do so, but generally we'll just end up ignoring certs that can't possibly be used to chain build.

**Comment by ptoffy** [comment]
Shouldn't the untrusted chain _not_ contain the root by definition? I feel like the place for the root should really just be the trusted store and it might be conceptually wrong to leave it in?

**Comment by Lukasa** [comment]
Generally speaking it's not helpful to include the root in the untrusted chain, we'll never be able to chain build through it. However, the only way for you to know whether that's a root, or a different relevant cert, is to try to chain build. It's probably better for us to do that than for you to.

So long as the chain isn't unreasonably long, the extra cost of us hashing the subject name is very cheap, so it's probably not worth you trying to avoid it.

**Comment by ptoffy** [comment]
Cool, thanks
